src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/model/Parameterisation.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * pottery-backend-interface - Backend API for testing programming exercises
+ * Copyright © 2015-2018 BlueOptima Limited, Andrew Rice (acr31@cam.ac.uk)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package uk.ac.cam.cl.dtg.teaching.pottery.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.wordnik.swagger.annotations.ApiModelProperty;
+
+import java.util.List;
+import java.util.Map;
+
+/** A Parameterisation represents the instructions for parameterising a task. */
+public class Parameterisation {
+  @ApiModelProperty("Values that are used for parameterisation.")
+  private final Map<String, List<JsonNode>> values;
```

Comment:
why is this a list of JsonNodes and not just a list of strings? (have i forgotten something from the plan?)

Reply:
Basically, because we might want e.g. the textual representation and numeric representation of a value available. I have some values like:
```"numbers": [
                {"digit": 4, "name": "four", "d4": true},
                {"digit": 5, "name": "five", "d5": true},
...
            ],
```
Which I think are necessary because we wanted some simple templating. I'll send you an e-mail with the complete task.json for that one and hopefully it'll make sense.

Reply:
ok. my concern here is tying our internal code to the arbitrary file format we chose for task.json. If we wanted to use e.g. yaml in future this gets messy.  Options: use Object and punt the problem to later? come up with some type that expresses what options we are allowing? switch to some sort of function passing style where the parameterisation class provides ways to apply a parameter to something without exposing internal representation. 

Sorry - I'm not dragging my heels for the sake of it. It seems like things are getting very complicated and I wonder if that's because we've made a poor decision somewhere.

Reply:
Yeap, I see where you're coming from here. A suitable type from my perspective would be (in ML+Java-like syntax):
`type Option = StringOption String | MapOption Map<String, Option>`

I quickly knocked up something to do this; it's about 100 lines to do that including the custom Jackson deserialization code to generate it. Do you fancy that?

(I also like the function passing style, but this is probably simpler.)